### ISSUE TITLE
DevTools, Observatory, Xcode/Instruments disagree about total memory usage.

### ISSUE THREAD

**Issue by kangwang1988** [flutter/flutter#55934]
## Steps to Reproduce
Sorry, it's confidential code and we can't give a minimal demo for reproducing.

## Details

Xcode reports high memory usage for one of our application( @fydx is in charge of that application), while Observatory reports low memory use for flutter(dart) side.

### Background information

It's an application where flutter is using in add-to-app form.
In this scenario, the home page uses flutter, in other words, it's the first page after launching.
The page structure is given below(The snapshot is taken for an application from appstore)
![image](https://user-images.githubusercontent.com/817851/80560770-da059780-8a14-11ea-946d-9753c713e78f.png)

### Problem Description

![image](https://user-images.githubusercontent.com/817851/80559728-19ca8000-8a11-11ea-9161-fce7f618a75e.png)

P.S. 
All screenshots are taken where flutter is running in profile mode.

### More information

I did more investigation and noticed some confusing thing.
a. If the home page is an empty UIViewController, i.e., all the flutter environment is not launched.
Xcode reports about 45MB memory usage, let's call it a homepage baseline.
b. If the home page return a Text widget instead of it's designed business logic, xcode reports about 125MB memory usage. In other words, we can say that flutter engine and simplest flutter business logic brings about 80MB memory increment.
![image](https://user-images.githubusercontent.com/817851/80560404-62833880-8a13-11ea-8d20-8e67970819be.png)
This data(80MB) also matches the flutter_gallery one where I forced to return a Text widget.
c. When I removed test code to allow HomePage return what it's designed, the memory increases a lot. 
![image](https://user-images.githubusercontent.com/817851/80559728-19ca8000-8a11-11ea-9161-fce7f618a75e.png)
The xcode reports about 90MB of memory increment while Observatory reports only 30MB of memory increment. I think it is probably resulted from flutter related code, however, the observatory make me pretty confusing.

```
[✓] Flutter (Channel stable/v1.12.13+hotfix.8, v1.12.13+hotfix.8, on Mac OS X 10.15.4 19E287, locale en-CN)
    • Flutter version 1.12.13+hotfix.8 at /Users/kylewong/Codes/Flutter/alibaba-flutter/StableV1.12.13/official/flutter
    • Framework revision 0b8abb4724 (3 months ago), 2020-02-11 11:44:36 -0800
    • Engine revision e1e6ced81d
    • Dart version 2.7.0

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/kylewong/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-R, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode11.3.1.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C505
    ! Unknown CocoaPods version installed.
        Flutter is unable to determine the installed CocoaPods's version.
        Ensure that the output of 'pod --version' contains only digits and . to be recognized by Flutter.
      To upgrade:
        sudo gem install cocoapods

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.42.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.8.1
```

**Comment by kangwang1988** [comment]
@liyuqian @zoeyfan

**Comment by mraleph** [comment]
DevTools (/ Observatory) are only showing Dart allocations (which Dart VM knows precisely) and external allocations (which are approximately reported by Flutter Engine to the Dart VM) - the rest of native memory allocations are completely unknown to the Dart VM and thus are not reflected by the DevTools(/Observatory). 

Have you tried looking at the memory consumption using native XCode tools? I have not used these tools in a while, but I think you can export memory graph dump for your application and then run `vmmap --summary YourApp.memgraph` that should provide a very informative breakdown of the native memory usage. 

Could you try using that on your application and post the result?

**Comment by dnfield** [comment]
I normally use Xcode when analyzing memory usage, because I like its tools.

It is definitely true that native allocations may not be reflected in the observatory. In some cases we can probably improve that by returning better values from places like https://github.com/flutter/engine/blob/fa0e464a710790c6110d541ae022d70e2f9009d8/lib/ui/painting/image.cc#L43

However, my understanding is that the VM expects this to be set once and not changed (@rmacnak-google  or @mraleph would know more about that and if it's true/still true).

It would be helpful if you can identify any objects in Xcode that seem like they should have counterparts in the Dart Observatory but don't. One thing in particular where we cheat is around layers - we don't know the size of the layers when we make them, and so we just provide a heuristic that seems to make garbage collection work ok (https://github.com/flutter/engine/blob/f2dbeb8aa76c1bc28755c12328ead828947f60bc/lib/ui/painting/engine_layer.cc#L21-L26). I suspect that is a significant source of mismatches, but again we can't really fix it becuase once we set it we can't update it.

**Comment by kangwang1988** [comment]
What make me confused the most is that it seems high memory usage report is given when a real flutter page is launched, it reports low for an empty flutter widget. So I doubt is might be related to flutter, however, everything I got from observatory shows low memory use.

**Comment by kangwang1988** [comment]
@mraleph 
kylewong@KyleWongdeMacBook-Pro Desktop % vmmap --summary MemoryGraph.memgraph 
Hardware Model:  iPhone7,1
Process:         LTao4iPhone [3554]
Path:            /private/var/containers/Bundle/Application/DC6862E3-9FA3-4AE4-8A89-B50BFE4E9147/LTao4iPhone.app/LTao4iPhone
Load Address:    0x100af0000
Identifier:      LTao4iPhone
Version:         ???
Code Type:       ARM64
Parent Process:  debugserver [3553]

Date/Time:       2020-04-30 00:55:44.935 +0800
Launch Time:     2020-04-30 00:52:22.785 +0800
OS Version:      iPhone OS 12.4.5 (16G161)
Report Version:  104

Physical footprint:         244.5M
Physical footprint (peak):  269.3M
----

ReadOnly portion of Libraries: Total=561.1M resident=127.3M(23%) swapped_out_or_unallocated=433.8M(77%)
Writable regions: Total=849.3M written=109.7M(13%) resident=188.2M(22%) swapped_out=21.1M(2%) unallocated=640.1M(75%)

                                VIRTUAL RESIDENT    DIRTY  SWAPPED VOLATILE   NONVOL    EMPTY   REGION 
REGION TYPE                        SIZE     SIZE     SIZE     SIZE     SIZE     SIZE     SIZE    COUNT (non-coalesced) 
===========                     ======= ========    =====  ======= ========   ======    =====  ======= 
Activity Tracing                   256K      52K      52K       0K       0K      52K       0K        1 
CG image                            64K      16K      16K      40K       0K       0K       0K        2 
CG raster data                    17.6M     9.9M     9.9M    7764K       0K       0K       0K       10 
CoreAnimation                      512K     140K     140K     160K       0K     136K       0K       29 
CoreData Object IDs               4112K       4K       4K       0K       0K       0K       0K        2 
CoreUI image data                  112K      12K      12K      88K       0K       0K       0K        1 
CoreUI image file                  176K       0K       0K     160K       0K       0K       0K        4 
Foundation                         400K     384K     384K       0K       0K       0K       0K        2 
IOKit                             69.2M    66.6M    66.6M      64K       0K    65.8M       0K      110 
IOKit (reserved)                   328K       0K       0K       0K       0K       0K       0K       42         reserved VM address space (unallocated)
IOSurface                         30.3M    30.3M    30.3M       0K       0K    30.2M       0K        8 
IOSurface (reserved)                56K       0K       0K       0K       0K       0K       0K        5         reserved VM address space (unallocated)
Image IO                          6176K     748K     748K    1512K     112K       0K    3680K       39 
Kernel Alloc Once                   32K       8K       8K       0K       0K       0K       0K        1 
MALLOC guard page                  512K       0K       0K       0K       0K       0K       0K       32 
MALLOC metadata                    560K     144K     144K      68K       0K       0K       0K       35 
MALLOC_LARGE                      25.0M    9708K    9708K    15.3M       0K       0K       0K      199         see MALLOC ZONE table below
MALLOC_LARGE metadata               32K      32K      32K       0K       0K       0K       0K        1         see MALLOC ZONE table below
MALLOC_NANO                      512.0M       8K       8K       8K       0K       0K       0K        1         see MALLOC ZONE table below
MALLOC_SMALL                      64.0M    7688K    7304K     884K       0K       0K       0K        8         see MALLOC ZONE table below
MALLOC_TINY                       30.0M    15.6M    15.6M     668K       0K       0K       0K       30         see MALLOC ZONE table below
Mach message                       496K     124K     124K       0K       0K       0K       0K       31 
Memory Tag 26                     3824K       0K       0K       0K       0K       0K       0K        3 
Performance tool data             16.5M    16.5M    16.5M      24K       0K       0K       0K       15         not counted in TOTAL below
SQLite page cache                 1152K     572K     572K      16K     532K      40K     320K       18 
STACK GUARD                        688K       0K       0K       0K       0K       0K       0K       43 
Skywalk Networking                 256K     156K     156K       0K       0K     140K       0K       16 
Stack                             24.1M    1028K     996K     156K       0K       0K       0K       45 
Stack (reserved)                   544K       0K       0K       0K       0K       0K       0K        1         reserved VM address space (unallocated)
Stack Guard                         16K       0K       0K       0K       0K       0K       0K        1 
VM_ALLOCATE                       61.2M    38.7M    38.7M    2916K       0K       0K       0K       82 
VM_ALLOCATE (reserved)            2864K       0K       0K       0K       0K       0K       0K        2         reserved VM address space (unallocated)
WebKit Malloc                      160K       0K       0K     112K       0K       0K       0K        6 
__DATA                            23.1M    13.5M    9875K    2605K       0K       0K       0K      447 
__DATA_CONST                      46.9M    33.7M     702K    1149K       0K       0K       0K      461 
__DATA_DIRTY                      1086K     689K     491K      57K       0K       0K       0K      334 
__FONT_DATA                          4K       0K       0K       0K       0K       0K       0K        1 
__GLSLBUILTINS                    1904K       0K       0K       0K       0K       0K       0K        1 
__LINKEDIT                       149.5M    29.2M       0K       0K       0K       0K       0K       19 
__TEXT                           411.7M    98.2M      60K       4K       0K       0K       0K      470 
__UNICODE                          564K     292K       0K       0K       0K       0K       0K        1 
libnetwork                         640K       4K       4K       0K       0K       4K     624K       16 
mapped file                      118.8M    4820K      16K       0K       0K       0K       0K       35 
shared memory                     1440K     764K     764K     600K       0K     224K       0K       15 
unused but dirty shlib __DATA      168K     168K     168K       0K       0K       0K       0K      163 
===========                     ======= ========    =====  ======= ========   ======    =====  ======= 
TOTAL                              1.6G   362.5M   192.8M    33.9M     644K    96.7M    4624K     2773 
TOTAL, minus reserved VM space     1.6G   362.5M   192.8M    33.9M     644K    96.7M    4624K     2773 

                                           VIRTUAL   RESIDENT      DIRTY    SWAPPED ALLOCATION      BYTES DIRTY+SWAP          REGION
MALLOC ZONE                                   SIZE       SIZE       SIZE       SIZE      COUNT  ALLOCATED  FRAG SIZE  % FRAG   COUNT
===========                                =======  =========  =========  =========  =========  =========  =========  ======  ======
DefaultMallocZone_0x108770000               512.0M         8K         8K         8K          2         64        16K    100%       1
MallocStackLoggingLiteZone_0x108778000       58.0M      32.2M      31.8M      16.7M     198027      46.5M      2081K      5%     219
GPUSupport_mallocZone_0x10838c000            18.0M       108K       108K        16K        782        65K        59K     48%       4
AGX_glrAGXUShaderMallocZone_0x1085c8000      18.0M        76K        76K        12K        105        48K        40K     46%       4
WebKit Using System Malloc_0x10a694000       10.0M         4K         4K        52K        329        22K        34K     62%       3
QuartzCore_0x10a6ec000                       10.0M       232K       232K        20K       2856       193K        59K     24%       3
NWMallocZone_0x10add0000                     2048K        20K        20K         8K         31         4K        24K     85%       2
AGX_glrUSCMallocZone_0x1085f0000             2048K         4K         4K        20K          5        288        24K     99%       2
MallocHelperZone_0x106d98000                 1024K         4K         4K         8K          1         32        12K    100%       1
===========                                =======  =========  =========  =========  =========  =========  =========  ======  ======
TOTAL                                       631.0M      32.7M      32.3M      16.8M     202138      46.8M      2349K      5%     239

**Comment by kangwang1988** [comment]
More information is given with instruments.
<img width="1440" alt="Screen Shot 2020-04-30 at 12 48 17 AM" src="https://user-images.githubusercontent.com/817851/80624186-ea059180-8a7d-11ea-843d-7580f0286022.png">

**Comment by liyuqian** [comment]
I suspect that the major difference in those memory measurements are caused by images, so I did some experiments using the sample app in https://github.com/flutter/flutter/issues/19558#issuecomment-557021269 against v1.12.13 Flutter (cf37c2cd07a1d3ba296efff2dc75e19ba65e1665) and tip-of-tree Flutter (a18e6361cf7b5b56ddc8c13fbc871dd10348e00c) with the following findings.

1. DevTools' memory measurements are always much smaller than Xcode instruments measurements (less than ~60 MB vs ~120 MB).
2. Observatory homepage's memory measurements are close to Xcode (~120 MB) in v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665, but much less than Xcode in tip-of-tree a18e6361cf7b5b56ddc8c13fbc871dd10348e00c (~60 MB vs ~120 MB).
3. Observatory's allocation profile page always gave a sum of memory measurements close to Xcode.

See the end of this comment for screenshots of (1) DevTools, (2) Observatory homepage, (3) Observatory allocation profile, and (4) Xcode measurements in (A) v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665 and (B) tip-of-tree a18e6361cf7b5b56ddc8c13fbc871dd10348e00c.

Based on those findings, I suggest
- @kangwang1988 : Can you please try to see if the Observatory gave close-to-Xcode memory measurements in v1.12.13 for your app? (The screenshot you gave is from DevTools instead of Observatory. )
- @terrylucas : Can you please make DevTools' memory page report the same memory usage as the Observatory's allocation profile page?
- @mraleph : Can you please investigate why Observatory homepage has close-to-Xcode measurement in v1.12.13 but not in tip-of-tree? Is that an unexpected regression?
- @dnfield : Once you're back and have some free time, can you please please investigate if we can add a devicelab test to assert that DevTools gave similar readings as Xcode? https://github.com/flutter/flutter/issues/44013 and https://github.com/flutter/flutter/pull/55486 may be helpful for the DevTools side. @iskakaushik may be able to help for the Xcode side with his [new profiling work](https://docs.google.com/document/d/1z_V69XzRct3vR7mBZ9tuuS1pZ_e7FQkuzbVJjXxMeaI/edit#bookmark=id.gm0025wm1ccz).

Also CC @cobblest and @InMatrix for UXR awareness.

# (A) Screenshots for v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665

## (1A) DevTools for v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665
<img width="1637" alt="Screen Shot 2020-04-29 at 6 37 53 PM" src="https://user-images.githubusercontent.com/22987568/80672316-3abac180-8a61-11ea-9a39-a0e33aa4d9ee.png">

## (2A) Observatory homepage for v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665
<img width="1623" alt="Screen Shot 2020-04-29 at 6 36 18 PM" src="https://user-images.githubusercontent.com/22987568/80672326-3ee6df00-8a61-11ea-90f2-9737fd291f4c.png">

## (3A) Observatory allocation profile page for v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665
<img width="1556" alt="Screen Shot 2020-04-29 at 6 36 37 PM" src="https://user-images.githubusercontent.com/22987568/80672323-3e4e4880-8a61-11ea-8d5c-a7b8396a60b8.png">

## (4A) Xcode for v1.12.13 cf37c2cd07a1d3ba296efff2dc75e19ba65e1665
<img width="998" alt="Screen Shot 2020-04-29 at 6 37 17 PM" src="https://user-images.githubusercontent.com/22987568/80672322-3db5b200-8a61-11ea-821c-2bf66fc61001.png">

# (B) Screenshots for tip-of-tree a18e6361cf7b5b56ddc8c13fbc871dd10348e00c

## (1B) DevTools for tip-of-tree a18e6361cf7b5b56ddc8c13fbc871dd10348e00c
<img width="1634" alt="Screen Shot 2020-04-29 at 6 43 06 PM" src="https://user-images.githubusercontent.com/22987568/80672496-d5b39b80-8a61-11ea-84fc-8f5974020c7a.png">

## (2B) Observatory homepage for tip-of-tree a18e6361cf7b5b56ddc8c13fbc871dd10348e00c
<img width="1670" alt="Screen Shot 2020-04-29 at 6 41 21 PM" src="https://user-images.githubusercontent.com/22987568/80672503-da784f80-8a61-11ea-9878-407cd793bd21.png">

## (3B) Observatory allocation profile page for tip-of-tree a18e6361cf7b5b56ddc8c13fbc871dd10348e00c
<img width="1580" alt="Screen Shot 2020-04-29 at 6 41 36 PM" src="https://user-images.githubusercontent.com/22987568/80672501-d9472280-8a61-11ea-88e4-dbd09946edcc.png">

## (4B) Xcode for tip-of-tree <img width="1006" alt="Screen Shot 2020-04-29 at 6 42 24 PM" src="https://user-images.githubusercontent.com/22987568/80672498-d815f580-8a61-11ea-9ff4-2b335c2e8b96.png">

**Comment by kangwang1988** [comment]
Also cc @wqyfavor

**Comment by mraleph** [comment]
> @mraleph : Can you please investigate why Observatory homepage has close-to-Xcode measurement in v1.12.13 but not in tip-of-tree? Is that an unexpected regression?

@liyuqian I don't really have an iOS device available due to WFH situation (waiting for one to arrive from Apple Store though). I looked at the code and I don't think it has really changed - the underlying implementation should bottom out in `Process::CurrentRSS` and `Process::MaxRSS` [here](https://cs.opensource.google/dart/sdk/+/master:runtime/bin/process_macos.cc;l=963?q=Process::MaxRSS) and this code has not changed in many years. These numbers are different from other numbers in the Observatory because they are comming directly from OS itself - so I would expect them to match what XCode shows. 

If you have an iOS device you could put debug prints into these two methods and see what they report when you navigate to Observatory page.

**Comment by jacob314** [comment]
> It is definitely true that native allocations may not be reflected in the observatory. In some cases we can probably improve that by returning better values from places like https://github.com/flutter/engine/blob/fa0e464a710790c6110d541ae022d70e2f9009d8/lib/ui/painting/image.cc#L43
> 
> However, my understanding is that the VM expects this to be set once and not changed (@rmacnak-google or @mraleph would know more about that and if it's true/still true).
> 
> It would be helpful if you can identify any objects in Xcode that seem like they should have counterparts in the Dart Observatory but don't. One thing in particular where we cheat is around layers - we don't know the size of the layers when we make them, and so we just provide a heuristic that seems to make garbage collection work ok (https://github.com/flutter/engine/blob/f2dbeb8aa76c1bc28755c12328ead828947f60bc/lib/ui/painting/engine_layer.cc#L21-L26). I suspect that is a significant source of mismatches, but again we can't really fix it becuase once we set it we can't update it.

Can we get some data on how off the memory usage estimate is for the layer case? Being able to update the memory usage estimate for layers when computing a memory snapshot seems useful. Are there any other cases where the flutter engine is having to guess the native memory usage?

**Comment by Hixie** [comment]
Do we have a page on flutter.dev explaining how to use all the relevant tools to diagnose memory usage? We should update (or create) such a page to include all the information in this bug and anything else we can come up with. Doing that, plus providing tooling that clearly indicates where memory usage is going and how to reduce it, seems like the most critical thing we can do for both customer: gold and customer: money at this time.

**Comment by Hixie** [comment]
cc @cbracken @mehmetf

**Comment by cobblest** [comment]
current documentation page for memory profiler: https://flutter.dev/docs/development/tools/devtools/memory

**Comment by dnfield** [comment]
AFAIK, we do not have documentation on how to use native platform memory profiling tools, which unfortunately is very important at the moment if memory is your concern, since devtools cannot capture the full scope of your program's memory usage.

Also, unfortunately, the various platforms differ in quality for memory debugging.

**Comment by liyuqian** [comment]
@dnfield : we have a memory sync meeting next Wednesday. Shall we gather folks to come up with a coordinated effort to document our current memory profiling best practices, and plan some works to improve our tools?

**Comment by liyuqian** [comment]
@jacob314 : for the layer memory, we had some discussions in https://github.com/flutter/flutter/issues/31498. So far, I'm not very concerned about the `SkPicture` size as it's often referenced by a Dart `Picture` object and we can count the size there. The layer raster cache image might be the big one that we may be missing right now. But currently I believe only the `OpacityLayer` and `ImageFilterLayer` can generate layer raster cache so it's unlikely that we're missing a lot of memory usage there.

**Comment by Hixie** [comment]
Creating documentation along the lines of that described above, for each platform, should IMHO be the target for resolving this bug.

**Comment by Hixie** [comment]
What is the target milestone for this bug?

**Comment by InMatrix** [comment]
We're running a user study with both 1p and 3p customers who have dealt with memory issues to identify documentation requirements, among other things. We aim to have results by early June. @cobblest is driving the study.

**Comment by dnfield** [comment]
It's not clear to me why this is critical or a blocker - Xcode is definitely giving a more accurate report than the observatory, which has less information to work with. We should fix this, but it's hard to see what exactly is the blocker or critical.

**Comment by kangwang1988** [comment]
Update:
Previously, in order to reuse a already-existing image library in native side, for example (SDWebImage, Cache, CDN size, etc), we did something modification so that the dart side image logic would be redirected to native side implementation using an external image codec&provider. Now according to @xujim, when trying another implementation for this reuse, the memory reduces significantly to an acceptable level. So this issue is not a blocker now.
Maybe the main problem indicated in this issue is that developers can't have a clear way to determine where the problem could be while observatory only knows partial information. In other words. If observatory can give an overview of the major memory use for the whole application, indicating that it's the native side memory(not only the external memory which can be known by the observatory now) that results in large memory consumption, if will be more clear and we can take a deeper look at our native side logic.

**Comment by sanlittt** [comment]
@liyuqian 
I found a similar problem that AndroidStudio reports high memory usage than Observatory. 
In my case, when entering a Flutter page, The Observatory shows a total increase of 47M (including the used +23M and the external +24M), But the Android Profile shows a memory increase of 126M (Native + 72.9m, Graphics + 46.7m). We don't know which part of the usage is causing the overall memory to grow so much, nor can we optimize memory performance.

    [✓] Flutter (Channel alf_stable_v1.12.13, v1.12.13+hotfix.8-pre.19, on Mac OS X 10.15.5 19F101, locale zh-Hans-CN)
    
    [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
    [✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    [✓] Android Studio (version 3.6)
    [!] IntelliJ IDEA Ultimate Edition (version 2019.3)
        ✗ Flutter plugin not installed; this adds Flutter specific functionality.
        ✗ Dart plugin not installed; this adds Dart specific functionality.
    [!] IntelliJ IDEA Community Edition (version 2018.3.2)
        ✗ Flutter plugin not installed; this adds Flutter specific functionality.
        ✗ Dart plugin not installed; this adds Dart specific functionality.
    [!] VS Code (version 1.43.2)
        ✗ Flutter extension not installed; install from
          https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter
    [✓] Connected device (1 available)
    


The memory info：
1.before
<img width="1196" alt="1" src="https://user-images.githubusercontent.com/43398347/87298988-c78ffa00-c53d-11ea-9fc4-c75dfa0a8337.png">
<img width="1182" alt="2" src="https://user-images.githubusercontent.com/43398347/87299013-d2e32580-c53d-11ea-8549-c2b10366eb01.png">

2.after
![3](https://user-images.githubusercontent.com/43398347/87299055-e7bfb900-c53d-11ea-9bb9-53affccbfab1.png)
<img width="1184" alt="4" src="https://user-images.githubusercontent.com/43398347/87299067-edb59a00-c53d-11ea-8288-3f30e5f6183e.png">

**Comment by liyuqian** [comment]
@terrylucas @jacob314 @dnfield : according to our meeting today, DevTools now should have displayed RSS memory that's consistent with adb or Xcode total memory measurement as it's basically displaying what the operating system tells us for the process memory usage. Do you think that would be sufficient to close this issue?

@sanlittt: it's true that there will always be some memory that the operating system can detect but the Dart VM can't detect (e.g., some Skia objects that's completely independent of the Dart VM). That's why we can only guarantee that the total RSS memory reported by DevTools matches adb/Xcode, and we can't guarantee that Dart VM's allocation profile matches adb/Xcode's memory measurement. Can you please check if DevTools' RSS memory increase matches adb or Android Profile's memory increase?

**Comment by liyuqian** [comment]
Just chatted with @jacob314 offline and confirmed that RSS memory should match Xcode/adb measurement.

However, one has to use `pub global activate devtools` to get the latest DevTools version `0.8.0` to see that RSS memory. The default DevTools even in the tip-of-tree master branch Flutter is too old (version `0.2.5` in my local experiment). So don't use `flutter run` and press `v` to open an old DevTools for memory debugging yet. @jacob314 @terrylucas : is there a Github issue that tracks the upgrade of DevTools in Flutter?

On top of that, the latest DevTools seem to still have the following 2 problems (CC @terrylucas):
1. The RSS memory is displayed on the left, but its timeline is out of the y-axis range.
2. The Android memory displayed on the right (presumably from adb) seems to be off by 1000x? It's 240KB instead of 240MB.

<img width="1662" alt="Screen Shot 2020-07-27 at 1 35 00 PM" src="https://user-images.githubusercontent.com/22987568/88590132-0da18f00-d00f-11ea-9a6b-292e73d780ce.png">

**Comment by liyuqian** [comment]
Reassign to @terrylucas and @jacob314 as this needs DevTools fixes.

**Comment by terrylucas** [comment]
This [PR](https://github.com/flutter/devtools/commit/709fa720442af8d219dee391729c5bc4b6ea41a8) fixes ADB output displaying in KB instead of total bytes.  I'm working on a PR to fix the RSS scale so it appears.

**Comment by terrylucas** [comment]
RSS not appearing in the chart has been fixed and is available in DevTools Release 0.9.3
